internal/gcsx/inactive_timeout_reader.go

```diff
@@ -0,0 +1,245 @@
+// underlying GCS reader. It attempts to create the initial reader using the
+// provided parameters. If successful, it starts a background goroutine to monitor
+// for inactivity based on the specified timeout.
+//
+// If the timeout duration is zero, it returns (nil, ErrZeroInactivityTimeout) as a zero timeout
+// defeats the purpose of this wrapper.
+func NewInactiveTimeoutReader(ctx context.Context, bucket gcs.Bucket, object *gcs.MinObject, readHandle []byte, byteRange gcs.ByteRange, timeout time.Duration) (gcs.StorageReader, error) {
+	return NewInactiveTimeoutReaderWithClock(ctx, bucket, object, readHandle, byteRange, timeout, clock.RealClock{})
+}
+
+func NewInactiveTimeoutReaderWithClock(ctx context.Context, bucket gcs.Bucket, object *gcs.MinObject, readHandle []byte, byteRange gcs.ByteRange, timeout time.Duration, clock clock.Clock) (gcs.StorageReader, error) {
+	if timeout == 0 {
+		return nil, ErrZeroInactivityTimeout
+	}
+
+	itr := &inactiveTimeoutReader{
+		object:     object,
+		bucket:     bucket,
+		reqRange:   byteRange,
+		readHandle: readHandle,
+		mu:         locker.New("inactiveTimeoutReader: "+object.Name, func() {}),
+		isActive:   false,
+	}
+	itr.ctx, itr.cancel = context.WithCancel(ctx)
+
+	var err error
+	if itr.gcsReader, err = itr.createGCSReader(); err != nil {
+		return nil, err
+	}
+
+	// Start the background periodic routine.
+	go itr.monitor(clock, timeout)
+
+	return itr, nil
+}
+
+// createGCSReader is a helper method to create the underlined reader from itr.start + itr.seen offset.
+func (itr *inactiveTimeoutReader) createGCSReader() (gcs.StorageReader, error) {
+	reader, err := itr.bucket.NewReaderWithReadHandle(
+		itr.ctx,
+		&gcs.ReadObjectRequest{
+			Name:       itr.object.Name,
+			Generation: itr.object.Generation,
+			Range: &gcs.ByteRange{
+				Start: itr.reqRange.Start + itr.seen,
+				Limit: itr.reqRange.Limit,
+			},
+			ReadCompressed: itr.object.HasContentEncodingGzip(),
+			ReadHandle:     itr.readHandle,
+		})
+	if err != nil {
+		return nil, fmt.Errorf("NewReaderWithReadHandle: %w", err)
+	}
+	return reader, nil
+}
+
+// Read implements io.Reader interface.
+//
+// If the underlying reader has been closed due to inactivity, Read automatically
+// attempts to reconnect using the stored read handle and the correct offset
+// (start + bytes previously seen). If reconnection fails, the error is returned.
+//
+// Each successful Read call marks the reader as active, resetting the inactivity timer
+// in the background monitor. This method is thread-safe.
+//
+// Calling Read() after explicitly calling Close() is not supported and will
+// lead to undefined behavior.
+func (itr *inactiveTimeoutReader) Read(p []byte) (n int, err error) {
+	itr.mu.Lock()
+	defer itr.mu.Unlock()
+
+	itr.isActive = true
+
+	if itr.gcsReader == nil {
+		if itr.gcsReader, err = itr.createGCSReader(); err != nil {
+			return 0, err
+		}
+	}
+
+	n, err = itr.gcsReader.Read(p)
+	itr.seen += uint64(n)
```

Comment:
I understood if err!= nil then n will be zero but should we explicitly return before if error is not nit for better readability? Also what if error is EOF?

Reply:
Semantically, there will not be any change from the underlined error to this wrapper. So, all the error will be handled similar to the underlined error.

Returning before will required an extra check for err != EOF, as Don't want to add an extra error check in the Read operation, unless important.